File: sdk/ruby/lib/cordum/cap/progress.rb
# frozen_string_literal: true

module Cordum
  module Cap
    module ProgressHelper
      # Builds a job progress payload (serialized BusPacket bytes).
      def self.progress_payload(sender_id:, job_id:, step_id:, percent:, message:)
        progress = Cordum::Agent::V1::JobProgress.new(
          job_id: job_id,
          step_id: step_id,
          percent: percent,
          message: message,
          status: :JOB_STATUS_RUNNING
        )

        pkt = Cordum::Agent::V1::BusPacket.new(
          sender_id: sender_id,
          protocol_version: DEFAULT_PROTOCOL_VERSION,
          created_at: Google::Protobuf::Timestamp.new(seconds: Time.now.to_i, nanos: 0),
          job_progress: progress
        )

        Codec.marshal_deterministic(pkt)
      end

      # Builds a job cancel payload (serialized BusPacket bytes).
      def self.cancel_payload(sender_id:, job_id:, reason:, requested_by:)
        cancel = Cordum::Agent::V1::JobCancel.new(
          job_id: job_id,
          reason: reason,
          requested_by: requested_by
        )

        pkt = Cordum::Agent::V1::BusPacket.new(
          sender_id: sender_id,
          protocol_version: DEFAULT_PROTOCOL_VERSION,
          created_at: Google::Protobuf::Timestamp.new(seconds: Time.now.to_i, nanos: 0),
          job_cancel: cancel
        )

        Codec.marshal_deterministic(pkt)
      end

      # Publishes a progress payload to NATS.
      def self.emit_progress(nc, payload)
        nc.publish(SUBJECT_PROGRESS, payload)
      end

      # Publishes a cancel payload to NATS.
      def self.emit_cancel(nc, payload)
        nc.publish(SUBJECT_CANCEL, payload)
      end
    end
  end
end

File: sdk/dotnet/src/Cordum.Cap/Progress.cs
using Google.Protobuf.WellKnownTypes;
using NATS.Client.Core;
using Cordum.Agent.V1;

namespace Cordum.Cap;

/// <summary>
/// Progress and cancel payload constructors and emission helpers.
/// </summary>
public static class ProgressHelper
{
    /// <summary>
    /// Build a job progress payload.
    /// </summary>
    public static byte[] ProgressPayload(
        string senderId, string jobId, string stepId, int percent, string message)
    {
        var progress = new JobProgress
        {
            JobId = jobId,
            StepId = stepId,
            Percent = percent,
            Message = message,
        };

        var packet = new BusPacket
        {
            SenderId = senderId,
            ProtocolVersion = Subjects.DefaultProtocolVersion,
            CreatedAt = Timestamp.FromDateTimeOffset(DateTimeOffset.UtcNow),
            JobProgress = progress,
        };

        return Codec.MarshalDeterministic(packet);
    }

    /// <summary>
    /// Build a job cancel payload.
    /// </summary>
    public static byte[] CancelPayload(
        string senderId, string jobId, string reason, string requestedBy)
    {
        var cancel = new JobCancel
        {
            JobId = jobId,
            Reason = reason,
            RequestedBy = requestedBy,
        };

        var packet = new BusPacket
        {
            SenderId = senderId,
            ProtocolVersion = Subjects.DefaultProtocolVersion,
            CreatedAt = Timestamp.FromDateTimeOffset(DateTimeOffset.UtcNow),
            JobCancel = cancel,
        };

        return Codec.MarshalDeterministic(packet);
    }

    /// <summary>
    /// Publish a progress payload to the progress subject.
    /// </summary>
    public static async Task EmitProgress(NatsConnection nc, byte[] payload, CancellationToken ct = default)
    {
        await nc.PublishAsync(Subjects.Progress, payload, cancellationToken: ct);
    }

    /// <summary>
    /// Publish a cancel payload to the cancel subject.
    /// </summary>
    public static async Task EmitCancel(NatsConnection nc, byte[] payload, CancellationToken ct = default)
    {
        await nc.PublishAsync(Subjects.Cancel, payload, cancellationToken: ct);
    }
}
